### PR TITLE
[Enhancement] Refine garbage collection of cloud native table

### PR DESCRIPTION
**Pull request by sduzh**
- Remove GC module in BE, trigger GC regularly by FE
- Retain recently created versions
- Avoid deleting versions accessed by schema change and compaction tasks.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 14 / 130 (10.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionJob.java|0|7|00.00%|[39, 40, 41, 42, 113, 117, 121]|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|4|00.00%|[70, 71, 72, 74]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionScheduler.java|0|5|00.00%|[138, 147, 290, 303, 313]|
|:large_blue_circle:|com/starrocks/lake/vacuum/AutovacuumDaemon.java|0|96|00.00%|[50, 54, 55, 58, 59, 63, 64, 65, 66, 67, 68, 72, 74, 76, 79, 80, 81, 82, 83, 86, 88, 90, 92, 94, 97, 98, 99, 101, 102, 104, 105, 107, 108, 109, 110, 114, 116, 118, 124, 125, 126, 128, 130, 131, 132, 133, 134, 137, 140, 141, 142, 144, 145, 146, 147, 149, 150, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 164, 165, 166, 167, 168, 169, 170, 171, 173, 175, 176, 177, 178, 180, 181, 183, 184, 185, 186, 187, 188, 189, 191, 192, 193, 194, 195, 196]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|10|60.00%|[528, 532, 533, 536]|
|:large_blue_circle:|com/starrocks/common/Config.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/LakeTableSchemaChangeJob.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 13 / 127 (10.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionJob.java|0|7|00.00%|[39, 40, 41, 42, 113, 117, 121]|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|4|00.00%|[70, 71, 72, 74]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionScheduler.java|0|5|00.00%|[138, 147, 290, 303, 313]|
|:large_blue_circle:|com/starrocks/lake/vacuum/AutovacuumDaemon.java|0|94|00.00%|[49, 56, 57, 60, 61, 65, 66, 67, 68, 69, 73, 75, 77, 80, 81, 82, 83, 84, 87, 89, 90, 92, 94, 95, 96, 97, 98, 100, 103, 104, 105, 107, 108, 109, 110, 114, 116, 118, 124, 125, 126, 128, 130, 131, 132, 133, 134, 137, 140, 141, 142, 143, 145, 146, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 160, 161, 162, 163, 164, 165, 166, 167, 169, 171, 172, 173, 174, 176, 177, 179, 180, 181, 182, 183, 184, 185, 186, 187, 189, 190, 192, 193, 194]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|10|60.00%|[528, 532, 533, 536]|
|:large_blue_circle:|com/starrocks/common/Config.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/LakeTableSchemaChangeJob.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 14 / 130 (10.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionJob.java|0|7|00.00%|[39, 40, 41, 42, 113, 117, 121]|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|4|00.00%|[70, 71, 72, 74]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionScheduler.java|0|5|00.00%|[138, 147, 290, 303, 313]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[758, 1370]|
|:large_blue_circle:|com/starrocks/lake/vacuum/AutovacuumDaemon.java|0|94|00.00%|[49, 56, 57, 60, 61, 65, 66, 67, 68, 69, 73, 75, 77, 80, 81, 82, 83, 84, 87, 89, 90, 92, 94, 95, 96, 97, 98, 100, 103, 104, 105, 107, 108, 109, 110, 114, 116, 118, 124, 125, 126, 128, 130, 131, 132, 133, 134, 137, 140, 141, 142, 143, 145, 146, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 160, 161, 162, 163, 164, 165, 166, 167, 169, 171, 172, 173, 174, 176, 177, 179, 180, 181, 182, 183, 184, 185, 186, 187, 189, 190, 192, 193, 194]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|10|60.00%|[528, 532, 533, 536]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/LakeTableSchemaChangeJob.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 1]
```diff
@@ -2219,6 +2219,25 @@ public class Config extends ConfigBase {
...
     public static int experimental_lake_publish_version_threads = 16;
 
+    @ConfField(mutable = true)
+    public static int lake_autovacuum_max_previous_versions = 0;
+
+    @ConfField
+    public static int lake_autovacuum_parallel_partitions = 8;
+
+    @ConfField(mutable = true)
+    public static long lake_autovacuum_partition_naptime_seconds = 30;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/Config.java:2229]: a little small, this may cause gc too frequent?

Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java:2229]: Do you have any suggested value?

File: be/src/common/config.h [hunk 2]
```diff
@@ -835,9 +835,6 @@ CONF_Int32(starlet_fs_stream_buffer_size_bytes, "131072");
 
 CONF_mInt64(lake_metadata_cache_limit, /*2GB=*/"2147483648");
 CONF_mBool(lake_print_delete_log, "true");
-CONF_mInt64(lake_gc_metadata_max_versions, "10");
-CONF_mInt64(lake_gc_metadata_check_interval, /*30 minutes=*/"1800");
-CONF_mInt64(lake_gc_segment_check_interval, /*60 minutes=*/"3600");
```
Comment (wyb) [inline be/src/common/config.h:840]: some other configs are not used, such as lake_enable_aggressive_gc, lake_aggressive_gc_high_priority.

File: fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+public class AutovacuumDaemon extends Daemon {
+    private static final Logger LOG = LogManager.getLogger(AutovacuumDaemon.class);
+
+    private static final long MILLISECONDS_PER_SECOND = 1000;
+    private static final long SECONDS_PER_MINUTE = 60;
+    private static final long MINUTES_PER_HOUR = 60;
+    private static final long MILLISECONDS_PER_HOUR = MINUTES_PER_HOUR * SECONDS_PER_MINUTE * MILLISECONDS_PER_SECOND;
+
+    private final ConcurrentHashMap<Long, Boolean> vacuumingPartitions = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java:56]: No ConcurrentHashSet in Java 8

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java (hunk 1 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java]: I changed the default value to 180 seconds(3 minutes)

File: gensrc/proto/lake_service.proto [hunk 4]
```diff
@@ -182,6 +182,42 @@ message AbortCompactionResponse {
     optional StatusPB status = 1;
 }
 
+message VacuumRequest {
+    repeated int64 tablet_ids = 1;
+    // Tablet metadata files with version numbers greater than or equals to min_retain_version
+    // will NOT be vacuumed. For tablet metadata files with version numbers less than
+    // min_retaion_version, decide whether they should be deleted by comparing the create time
```
[the remark this answers is not in the document]
Comment (sduzh) [reply gensrc/proto/lake_service.proto]: Updated

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 14 / 130 (10.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionJob.java|0|7|00.00%|[39, 40, 41, 42, 113, 117, 121]|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|4|00.00%|[70, 71, 72, 74]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionScheduler.java|0|5|00.00%|[138, 147, 290, 303, 313]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[758, 1370]|
|:large_blue_circle:|com/starrocks/lake/vacuum/AutovacuumDaemon.java|0|94|00.00%|[49, 56, 57, 60, 61, 65, 66, 67, 68, 69, 73, 75, 77, 80, 81, 82, 83, 84, 87, 89, 90, 92, 94, 95, 96, 97, 98, 100, 103, 104, 105, 107, 108, 109, 110, 114, 116, 118, 124, 125, 126, 128, 130, 131, 132, 133, 134, 137, 140, 141, 142, 143, 145, 146, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 160, 161, 162, 163, 164, 165, 166, 167, 169, 171, 172, 173, 174, 176, 177, 179, 180, 181, 182, 183, 184, 185, 186, 187, 189, 190, 192, 193, 194]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|10|60.00%|[528, 532, 533, 536]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/LakeTableSchemaChangeJob.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java:56]: `Sets.newConcurrentHashSet()` in guava

[the remark this answers is not in the document]
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/lake/vacuum/AutovacuumDaemon.java]: Updated

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 14 / 128 (10.94%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionJob.java|0|7|00.00%|[39, 40, 41, 42, 113, 117, 121]|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|4|00.00%|[70, 71, 72, 74]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionScheduler.java|0|5|00.00%|[138, 147, 290, 303, 313]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[758, 1370]|
|:large_blue_circle:|com/starrocks/lake/vacuum/AutovacuumDaemon.java|0|92|00.00%|[50, 57, 58, 61, 62, 66, 67, 68, 69, 70, 74, 76, 78, 81, 82, 83, 84, 85, 88, 90, 91, 93, 95, 96, 97, 98, 99, 101, 104, 105, 106, 108, 109, 113, 115, 117, 123, 124, 125, 127, 129, 130, 131, 132, 133, 136, 139, 140, 141, 142, 144, 145, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 159, 160, 161, 162, 163, 164, 165, 166, 168, 170, 171, 172, 173, 175, 176, 178, 179, 180, 181, 182, 183, 184, 185, 186, 188, 189, 191, 192, 193]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|10|60.00%|[528, 532, 533, 536]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/LakeTableSchemaChangeJob.java|3|3|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Comment by sduzh** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 14 / 128 (10.94%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionJob.java|0|7|00.00%|[39, 40, 41, 42, 113, 117, 121]|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|4|00.00%|[70, 71, 72, 74]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionScheduler.java|0|5|00.00%|[138, 147, 290, 303, 313]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[758, 1370]|
|:large_blue_circle:|com/starrocks/lake/vacuum/AutovacuumDaemon.java|0|92|00.00%|[50, 57, 58, 61, 62, 66, 67, 68, 69, 70, 74, 76, 78, 81, 82, 83, 84, 85, 88, 90, 91, 93, 95, 96, 97, 98, 99, 101, 104, 105, 106, 108, 109, 113, 115, 117, 123, 124, 125, 127, 129, 130, 131, 132, 133, 136, 139, 140, 141, 142, 144, 145, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 159, 160, 161, 162, 163, 164, 165, 166, 168, 170, 171, 172, 173, 175, 176, 178, 179, 180, 181, 182, 183, 184, 185, 186, 188, 189, 191, 192, 193]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|10|60.00%|[528, 532, 533, 536]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/LakeTableSchemaChangeJob.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**